### PR TITLE
Add denote test #11495

### PR DESCRIPTION
**Pull request by SongpengLiu**
Add denotation, users now can use annotation to denote a test:
@pytest.mark.test
def hello():
assert 5 == 5

test case could be found in testing/test_denotation.py
the relative issue url is: https://github.com/pytest-dev/pytest/issues/11495

### CODE REVIEW

**Review by nicoddemus — CHANGES_REQUESTED** [review]
We need first decide how we want to mark tests.

I like the idea, I also think being able to mark tests explicitly and use short names is a good one, similar to how it is done in Rust.

I would prefer a `@pytest.test` decorator rather than a mark: a test is certainly a first class citizen, as fixtures are, so we should use `@pytest.test` directly.

The implementation for the decorator should be trivial: it can just add a `__pytest_test__` attribute to the function, and we can then check for this attribute during collection to identify test functions.

One downside is that there would be now two ways to declare a test, using the decorator or the traditional prefix, but I don't see this as a major problem.

We will also need to update the docs accordingly, showing the new-style, the rationale, and that the traditional style of using a `test_` prefix is never going anywhere.

---

Other than that, we also need buy-in from other maintainers, given this is an important new feature.

cc @The-Compiler @bluetech @RonnyPfannschmidt @Pierre-Sassoulas @asottile

**Comment by nicoddemus** [comment]
Other than the above, we also need buy-in from other maintainers, given this is an important new feature.

cc @The-Compiler @bluetech @RonnyPfannschmidt @Pierre-Sassoulas @asottile

**Review by Pierre-Sassoulas — COMMENTED** [review]
> I would prefer a @pytest.test decorator rather than a mark: a test is certainly a first class citizen, as fixtures are, so we should use @pytest.test directly.

Also otherwise it's a "breaking change", the ``test`` mark would now be reserved by pytest.

I'm not strongly against this, but it does create two ways to do thing. Also the problem from the initial issue (``can lead to regular functions being mistakenly identified as tests``) is not going to be fixed by adding another way to denote test because the ``test_``  based discovery is not going anywhere. Regarding ``the test outputs contain a lot of fluff making it harder to read``, we could remove the ``test_``prefix from pytest's output instead.

**Comment by asottile** [comment]
I don't really like this idea.  it's one of the things that really bothered me about nose that every test needed a special marker to be run and I fear that if we add this people will assume it's _required_ and sprinkle it everywhere

**Comment by nicoddemus** [comment]
> Also otherwise it's a "breaking change", the test mark would now be reserved by pytest.

Not really, it won't be reserved in any sense, if the user does not import it, it won't have any effect.

> I'm not strongly against this, but it does create two ways to do thing. 

Fair point, that's the downside.

> Also the problem from the initial issue (can lead to regular functions being mistakenly identified as tests) is not going to be fixed by adding another way to denote test because the test_ based discovery is not going anywhere. 

True, but I guess it can "fix" that issue (if one considers that a problem) in a particular code base.

> Regarding the test outputs contain a lot of fluff making it harder to read, we could remove the test_prefix from pytest's output instead.

That I would be 👎, seems like this would cause confusion.

> I don't really like this idea. it's one of the things that really bothered me about nose that every test needed a special marker to be run and I fear that if we add this people will assume it's required and sprinkle it everywhere

Fair point.

---

Let's see what others think, but indeed "two ways of doing the same thing" is somewhat problematic, specially for new comers, training material, having to explain the differences, etc.

**Comment by Pierre-Sassoulas** [comment]
>>  (can lead to regular functions being mistakenly identified as tests) is not going to be fixed by adding another way to denote test because the test_ based discovery is not going anywhere.

>  I guess it can "fix" that issue (if one considers that a problem) in a particular code base.

The proposed change permit to denote that a function with a name not starting with ``test_`` is in fact a test, but not that a function with a name starting with ``test_`` is NOT a test (original issue from https://github.com/pytest-dev/pytest/issues/11494). Or did I miss something ?

File: src/_pytest/mark/structures.py [hunk 1]
```diff
@@ -616,3 +616,69 @@ def __len__(self) -> int:
 
     def __repr__(self) -> str:
         return f"<NodeKeywords for node {self.node}>"
+
+
+@final
+@dataclasses.dataclass(frozen=True)
+class Test:
```
Comment (RonnyPfannschmidt) [inline src/_pytest/mark/structures.py:623]: Why create an copy of the mark class?

**Review by RonnyPfannschmidt — COMMENTED** [review]
How instead of a obnoxious `pytest.test` using a name like force-collect/conditional-collect

I'm currently not sure how to make this one fit nicely

**Comment by SongpengLiu** [comment]
Nicoddemus, thanks for your response. I think with this feature, we will have many further works to do. The first thing is to decide the mark's form with other maintainers.

**Comment by The-Compiler** [comment]
I'm also -1 on this, mainly for two reasons:

- I'm not aware of anything else in the Python space doing things that way (notably, unittest doesn't!). Thus, pytest would introduce a "new" way of marking tests, yet obviously not recommend it as the main way of doing that. Sounds like a recipe for confusion to me.
- If people want to have this is the main/only way of denoting tests (which was the original motivation in #11495!), it means that forgetting a decorator is a silent failure (your test just doesn't run). From my experience with a [rather exotic C++ test framework](https://www.cute-test.com/), anything that requires you to manually register tests leads to forgetting to do that. It's much easier to follow a `def test_...` naming convention than it is to remember putting `@pytest.test` (or whatever) on every test. The only way out of this is IDE support, where the IDE somehow figures this is a test (e.g. based on a naming convention... :upside_down_face:) and then alerts you that you probably forgot said decorator. However, as already mentioned above, this is a new and very exotic way of doing things — thus, buy-in from IDEs and editors is unlikely to happen. And if it does, we need to be careful that people don't suddenly think this is the main/proper way of denoting pytest tests...

So if we want this at all, I agree with @RonnyPfannschmidt that we need some sort of name that communicates that this isn't the main way of doing things. I indeed like `@pytest.force_collect` or `@pytest.force_test` or somesuch. But personally, I'd rather not have this at all.

**Comment by nicoddemus** [comment]
I'm convinced and change my vote to 👎 too.

**Comment by nicoddemus** [comment]
I'm closing this then, and adding a note to the original issue. 

Thanks everyone, specially @SongpengLiu for willing to contribute in the first place, sorry in the end this was not accepted. 👍

**Pull request closed without merging**